EngineLayer/GlycoSearch/GlycoSearchEngine.cs

```diff
@@ -0,0 +1,508 @@
+
+
+            if (!isOGlycoSearch)
+            {
+                Glycans = GlycanDatabase.LoadGlycan(GlobalVariables.GlycanLocations.Where(p=> System.IO.Path.GetFileName(p)==_glycanDatabase).First(), !isOGlycoSearch).OrderBy(p => p.Mass).ToArray();
+                //TO THINK: Glycan Decoy database.
+                //DecoyGlycans = Glycan.BuildTargetDecoyGlycans(NGlycans);
+
+            }
+            else
+            {
+                GlycanBox.GlobalOGlycans = GlycanDatabase.LoadGlycan(GlobalVariables.GlycanLocations.Where(p => System.IO.Path.GetFileName(p) == _glycanDatabase).First(), !isOGlycoSearch).ToArray();
+                GlycanBox.GlobalOGlycanModifications = GlycanBox.BuildGlobalOGlycanModifications(GlycanBox.GlobalOGlycans);
+                GlycanBox.OGlycanBoxes = GlycanBox.BuildOGlycanBoxes(_maxOGlycanNum, false).OrderBy(p => p.Mass).ToArray();
+            }
+        }
+
+        private Glycan[] Glycans { get; }
+        //private Glycan[] DecoyGlycans { get; }
+
+        protected override MetaMorpheusEngineResults RunSpecific()
+        {
+            double progress = 0;
+            int oldPercentProgress = 0;
+            ReportProgress(new ProgressEventArgs(oldPercentProgress, "Performing crosslink search... " + CurrentPartition + "/" + CommonParameters.TotalPartitions, NestedIds));
+
+            byte byteScoreCutoff = (byte)CommonParameters.ScoreCutoff;
+
+            int maxThreadsPerFile = CommonParameters.MaxThreadsToUsePerFile;
+            int[] threads = Enumerable.Range(0, maxThreadsPerFile).ToArray();
+            Parallel.ForEach(threads, (scanIndex) =>
+            {
+                byte[] scoringTable = new byte[PeptideIndex.Count];
+                List<int> idsOfPeptidesPossiblyObserved = new List<int>();
+
+                byte[] secondScoringTable = new byte[PeptideIndex.Count];
+                List<int> childIdsOfPeptidesPossiblyObserved = new List<int>();
+
+                List<int> idsOfPeptidesTopN = new List<int>();
+                byte scoreAtTopN = 0;
+                int peptideCount = 0;
+
+                for (; scanIndex < ListOfSortedMs2Scans.Length; scanIndex += maxThreadsPerFile)
+                {
+                    // Stop loop if canceled
+                    if (GlobalVariables.StopLoops) { return; }
+
+                    // empty the scoring table to score the new scan (conserves memory compared to allocating a new array)
+                    Array.Clear(scoringTable, 0, scoringTable.Length);
+                    idsOfPeptidesPossiblyObserved.Clear();
+                    idsOfPeptidesTopN.Clear();
+
+                    var scan = ListOfSortedMs2Scans[scanIndex];
+
+                    // get fragment bins for this scan
+                    List<int> allBinsToSearch = GetBinsToSearch(scan, FragmentIndex, CommonParameters.DissociationType);
+                    List<int> childBinsToSearch = null;
+
+                    //TO DO: limit the high bound limitation
```

Comment:
The funcion IndexedScoring use the parameter 'highestMassPeptideToLookFor', which I now use Double.PositiveInfinity. I assume if I set this up to the precursor, this may speed up the search. 
The whole 'IndexedScoring' is a 'Deprecated' one now. So I didn't do anything but keep the reminder.

Reply:
I don't quite understand. Isn't the "highestMassPeptideToLookFor" used to determine the array length? It can't be set at infinity? Doesn't it also need to be the same for the entire search, not just for each individual peptide, such that you can't specify it to be the precursor? Or am I thinking of something completely different?

Reply:
highestMassPeptideToLookFor help for limit the adding of ScoreTable. When peptide are larger than precursor, there is no need to score them in the index. 
I just tested that we can actually specify it to be a double larger than the current precursor. The result changed due to the top 50 candidate changed after ion-indexing, but not in a bad way.
Thanks for pointing this out.